swift/ql/lib/codeql/swift/printast/PrintAst.qll

```diff
@@ -0,0 +1,131 @@
+/**
+ * Provides queries to pretty-print a Go AST as a graph.
+ */
+
+import swift
+import codeql.swift.generated.ParentChild
+
+/**
+ * Hook to customize the files and functions printed by this module.
+ *
+ * For an AstNode to be printed, it always requires `shouldPrintFile(f)` to hold
+ * for its containing file `f`, and additionally requires `shouldPrintFunction(fun)`
+ * to hold if it is, or is a child of, function `fun`.
+ */
+class PrintAstConfiguration extends string {
+  /**
+   * Restrict to a single string, making this a singleton type.
+   */
+  PrintAstConfiguration() { this = "PrintAstConfiguration" }
+
+  /**
+   * Holds if the AST for `e` should be printed. By default, holds for all.
+   */
+  predicate shouldPrint(Locatable e) { any() }
+}
+
+private predicate shouldPrint(Locatable e) { any(PrintAstConfiguration config).shouldPrint(e) }
+
+/**
+ * An AST node that should be printed.
+ */
+private newtype TPrintAstNode =
+  TLocatable(Locatable ast) {
+    // Only consider resolved nodes (that is not within the hidden conversion AST)
+    ast = ast.resolve()
```

Comment:
Hmm. Shouldn't this be:
```ql
ast = any(Locatable loc).resolve()
```
?

That is, I think _any_ element targeted by a call to `resolve` should be a possible PrintAST node.

Reply:
Hmm, I think they are equivalent, because resolve is idempotent.

Reply:
Ah, right!